### ISSUE TITLE
Unicode in python sources

### ISSUE THREAD

**Issue by matthen** [thought-machine/please#322]
I am finding that unicode in python source files gives an error in the generated pex. It works fine on mac, but errors on linux.

If I do the following:

```
diff --git a/test/python_rules/BUILD b/test/python_rules/BUILD
index 85d51e2..0471056 100644
--- a/test/python_rules/BUILD
+++ b/test/python_rules/BUILD
@@ -122,3 +122,9 @@
     ],
     deps = ["//third_party/python:tensorflow"],
 )
+
+python_test(
+    name = "unicode_test",
+    srcs = ["unicode_test.py"],
+    labels = ["py3"],
+)
diff --git a/test/python_rules/unicode_test.py b/test/python_rules/unicode_test.py
new file mode 100644
index 0000000..49042e2
--- /dev/null
+++ b/test/python_rules/unicode_test.py
@@ -0,0 +1,9 @@
+# -*- coding: utf-8 -*-
+"""A test that please can run python with unicode in the source."""
+
+import unittest
+
+
+class UnicodeTest(unittest.TestCase):
+    def test_unicode(self):
+        self.assertEqual(u'kérem', u'kérem')
```

then run the test on debian:

```
docker run -it -v $(pwd):/please python:3.6 bash
cd
wget https://storage.googleapis.com/golang/go1.8.3.linux-amd64.tar.gz
tar -xzf go1.8.3.linux-amd64.tar.gz
mv go/ /usr/local
export GOROOT=/usr/local/go
export GOPATH=$HOME/work/goproj
ln -s /usr/local/go/bin/go /bin/go
apt-get update
apt-get install unzip


cd /please
./tools/misc/pleasew test test/python_rules:unicode_test
```

Then I get the following error:

```
Traceback (most recent call last):
  File "/please/plz-out/tmp/test/python_rules/unicode_test._test/unicode_test.pex/__main__.py", line 273, in import_tests
  File "/usr/local/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'test.python_rules.unicode_test'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/please/plz-out/tmp/test/python_rules/unicode_test._test/unicode_test.pex/__main__.py", line 315, in <module>
  File "/please/plz-out/tmp/test/python_rules/unicode_test._test/unicode_test.pex/__main__.py", line 307, in run
  File "/please/plz-out/tmp/test/python_rules/unicode_test._test/unicode_test.pex/__main__.py", line 177, in interact
  File "/please/plz-out/tmp/test/python_rules/unicode_test._test/unicode_test.pex/__main__.py", line 240, in main
  File "/please/plz-out/tmp/test/python_rules/unicode_test._test/unicode_test.pex/__main__.py", line 290, in run_tests
  File "/usr/local/lib/python3.6/unittest/suite.py", line 24, in __init__
    self.addTests(tests)
  File "/usr/local/lib/python3.6/unittest/suite.py", line 57, in addTests
    for test in tests:
  File "/please/plz-out/tmp/test/python_rules/unicode_test._test/unicode_test.pex/__main__.py", line 289, in <genexpr>
  File "/please/plz-out/tmp/test/python_rules/unicode_test._test/unicode_test.pex/__main__.py", line 276, in import_tests
  File "/usr/local/lib/python3.6/imp.py", line 235, in load_module
    return load_source(name, filename, file)
  File "/usr/local/lib/python3.6/imp.py", line 172, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 684, in _load
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 674, in exec_module
  File "<frozen importlib._bootstrap_external>", line 780, in get_code
  File "/usr/local/lib/python3.6/imp.py", line 156, in get_data
    return file.read()
  File "/usr/local/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 137: ordinal not in range(128)
```

(there is no error on my mac)

**Comment by peterebden** [comment]
Hmmmm, this works on my machine which also has python 3.6 - maybe a Ubuntu vs. Debian thing. I will have a bit of a dig with the docker example you provided.

(Sorry for the slow response, been a busy week with my real work...)

**Comment by peterebden** [comment]
This definitely seems to be some kind of property of the `python:3.6` image; it doesn't fail in e.g. `thoughtmachine/please_ubuntu`. I'm still drawing a blank as to what that would be though (it's not e.g. `sys.getdefaultencoding()` or `sys.getfilesystemencoding()`)...

**Comment by matthen** [comment]
Any more ideas?
Or how can I rewrite the code so it works, only using ascii? e.g. what if i want to do:
`if x == "é":...`

**Comment by BNMetrics** [comment]
@matthen 
Hello, This issue is caused by `imp` library we used, and it's fixed after [#356](https://github.com/thought-machine/please/issues/356).

I suspect this is because of `imp` being deprecated and unable to handle unicode characters.
